listenbrainz/troi/weekly_playlists.py

```diff
@@ -0,0 +1,177 @@
+        results = conn.execute(text(query), {"user_ids": user_ids, "export_preference": SPOTIFY_EXPORT_PREFERENCE})
+        for r in results:
+            details[r.user_id] = {
+                "username": r.musicbrainz_id,
+                "export_to_spotify": r.export_to_spotify
+            }
+
+            if r.export_to_spotify:
+                users_for_urls.append(r.user_id)
+
+    existing_urls = get_existing_playlist_urls(users_for_urls, slug)
+    for user_id, detail in details.items():
+        detail["existing_url"] = existing_urls.get(user_id)
+
+    return details
+
+
+def export_to_spotify(slug, description, playlists):
+    """ Export the playlists to spotify.
+
+        If a playlist url for the given user and slug already exists, its updated otherwise a new one is created.
+    """
+    service = SpotifyService()
+    for playlist in playlists:
+        try:
+            user_id = playlist["user_id"]
+            user = service.get_user(user_id, refresh=True)
+            if user is None:
+                current_app.logger.error("Unable to fetch spotify details for user_id: %d", user_id)
+                continue
+
+            sp = Spotify(auth=user["access_token"])
+
+            recordings = [Recording(mbid=mbid) for mbid in playlist["recordings"]]
+            playlist_element = Playlist(name=playlist["name"], description=description, recordings=recordings)
+            playlist_url, _ = submit_to_spotify(sp, playlist_element, user["external_user_id"],
+                                                existing_url=playlist["existing_url"])
+            playlist["additional_metadata"].update({"external_urls": {"spotify": playlist_url}})
+        except Exception:
+            current_app.logger.error("Unable to export playlist to spotify:", exc_info=True)
+
+
+def insert_recordings(cursor, playlist_id: str, recordings: list):
+    query = "INSERT INTO playlist.playlist_recording (playlist_id, position, mbid, added_by_id) VALUES %s"
+    template = SQL("({playlist_id}, %s, %s, {added_by_id})").format(
+        playlist_id=Literal(playlist_id),
+        added_by_id=Literal(TROI_BOT_USER_ID)
+    )
+    values = list(enumerate(recordings))
+    execute_values(cursor, query, values, template)
+
+
+def insert_playlists(cursor, playlists, description: str):
+    query = """
+        INSERT INTO playlist.playlist (creator_id, name, description, public, created_for_id, additional_metadata)
+             VALUES %s
+          RETURNING created_for_id, id
+    """
+    template = SQL("""({creator_id}, %s, {description}, 't', %s, %s)""").format(
+        creator_id=Literal(TROI_BOT_USER_ID),
+        description=Literal(description)
+    )
+    values = [(p["name"], p["user_id"], json.dumps(p["additional_metadata"])) for p in playlists]
+    results = execute_values(cursor, query, values, template, fetch=True)
+    return {r[0]: r[1] for r in results}
+
+
+def filter_and_update_playlists(slug, jam_name, all_playlists):
```

Comment:
It took me quite a while to work out that the playlist are not being deleted anywhere, but are being removed from the all_playlists list. Perhaps this function should be renamed? exclude_playlists_from_deleted_users?

Reply:
It excludes playlists and also updates the metadata for the remaining but I guess I could split it into two.

Reply:
I didn't mean for it to be split, just change the name to be more representative and all good.

Reply:
Makes sense